File: tray-app/MailMCP/ViewModels/AccountsViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
public final class AccountsViewModel {
    public private(set) var accounts: [AccountListItem] = []
    public private(set) var isLoading = false
    public private(set) var lastError: String?

    private let client: IpcClient
    private var notifTask: Task<Void, Never>?

    public init(client: IpcClient) {
        self.client = client
    }

    public func start() {
        Task { await refresh() }
        notifTask = Task { [weak self] in await self?.notificationLoop() }
    }

    public func stop() { notifTask?.cancel() }

    public func refresh() async {
        isLoading = true
        defer { isLoading = false }
        do {
            accounts = try await client.call("accounts.list")
            lastError = nil
        } catch {
            lastError = error.localizedDescription
        }
    }

    public func remove(id: AccountID) async {
        do {
            let _: Empty = try await client.call(
                "accounts.remove",
                params: ["account_id": .string(id.rawValue)]
            )
            await refresh()
        } catch {
            lastError = error.localizedDescription
        }
    }

    private func notificationLoop() async {
        do {
            let stream = try await client.subscribe(events: [
                "account.added", "account.removed", "account.needs_reauth",
            ])
            for await _ in stream { await refresh() }
        } catch {
            lastError = error.localizedDescription
        }
    }
}

File: tray-app-win/MailMCP/ViewModels/AccountsViewModel.cs
using System.Collections.ObjectModel;
using CommunityToolkit.Mvvm.ComponentModel;
using MailMCP.IPC;

namespace MailMCP.ViewModels;

/// <summary>
/// Live account list view-model. Fetches via <c>accounts.list</c> on start
/// and refreshes on every <c>account.added</c> / <c>account.removed</c> /
/// <c>account.needs_reauth</c> notification. Mirrors the v0.1b
/// AccountsViewModel.
/// </summary>
public partial class AccountsViewModel : ObservableObject
{
    private readonly IpcClient _client;
    private CancellationTokenSource? _cts;
    private Task? _notifTask;

    public ObservableCollection<AccountListItem> Accounts { get; } = new();

    [ObservableProperty] private bool _isLoading;
    [ObservableProperty] private string? _lastError;

    public AccountsViewModel(IpcClient client) { _client = client; }

    public void Start()
    {
        _cts = new CancellationTokenSource();
        _ = RefreshAsync(_cts.Token);
        _notifTask = Task.Run(() => NotificationLoop(_cts.Token));
    }

    public async Task StopAsync()
    {
        _cts?.Cancel();
        try
        {
            if (_notifTask is not null) await _notifTask.ConfigureAwait(false);
        }
        catch (OperationCanceledException) { }
    }

    public async Task RefreshAsync(CancellationToken ct = default)
    {
        IsLoading = true;
        try
        {
            var fresh = await _client.CallAsync<AccountListItem[]>("accounts.list", ct: ct)
                .ConfigureAwait(false);
            Accounts.Clear();
            foreach (var a in fresh) Accounts.Add(a);
            LastError = null;
        }
        catch (Exception ex)
        {
            LastError = ex.Message;
        }
        finally { IsLoading = false; }
    }

    public async Task RemoveAsync(string accountId, CancellationToken ct = default)
    {
        try
        {
            _ = await _client.CallAsync("accounts.remove", new { account_id = accountId }, ct)
                .ConfigureAwait(false);
            await RefreshAsync(ct).ConfigureAwait(false);
        }
        catch (Exception ex) { LastError = ex.Message; }
    }

    private async Task NotificationLoop(CancellationToken ct)
    {
        try
        {
            var stream = await _client.SubscribeAsync(
                new[] { "account.added", "account.removed", "account.needs_reauth" }, ct)
                .ConfigureAwait(false);
            await foreach (var note in stream.WithCancellation(ct).ConfigureAwait(false))
            {
                _ = note;  // any of these means refresh
                await RefreshAsync(ct).ConfigureAwait(false);
            }
        }
        catch (OperationCanceledException) { /* graceful */ }
        catch (Exception ex) { LastError = ex.Message; }
    }
}
